### PR TITLE
virtio-rng: add new cases for virtio-rng test

### PR DESCRIPTION
**Pull request by ybduan**
Migrate guest after hotplug/hotunplug rng device
ID: 1455802, 1455803
Signed-off-by: ybduan <yduan@redhat.com>

### CODE REVIEW

**Comment by ybduan** [comment]
python ConfigTest.py --testcase=rng_hotplug --guestname=RHEL.7.5 --rng_type=rng_random --clone=no

 (1/9) Host_RHEL.m7.u6.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.rng_hotplug.multi_rngs: PASS (108.23 s)
 (2/9) Host_RHEL.m7.u6.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.rng_hotplug.one_rng: PASS (76.75 s)
 (3/9) Host_RHEL.m7.u6.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.rng_hotplug.one_rng.with_reboot: PASS (150.64 s)
 (4/9) Host_RHEL.m7.u6.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.rng_hotplug.one_rng.with_system_reset: PASS (160.50 s)
 (5/9) Host_RHEL.m7.u6.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.rng_hotplug.one_rng.with_shutdown_after_unplug: PASS (75.63 s)
 (6/9) Host_RHEL.m7.u6.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.rng_hotplug.one_rng.with_shutdown_after_plug: PASS (64.54 s)
 (7/9) Host_RHEL.m7.u6.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.rng_hotplug.one_rng.with_migrate_after_unplug: PASS (139.44 s)
 (8/9) Host_RHEL.m7.u6.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.rng_hotplug.one_rng.with_migrate_after_plug: PASS (140.09 s)
 (9/9) Host_RHEL.m7.u6.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.rng_hotplug.repeat_in_loop: PASS (2147.24 s)

File: qemu/tests/rng_hotplug.py [hunk 1]
```diff
@@ -122,28 +130,36 @@ def run_subtest(sub_test):
...
+            # Unplug attached rng device
+            unplug_all_rngs(vm)
+            # temporary workaround for migration
+            vm.params["virtio_rngs"] = ''
+        else:
+            stop_rngd(vm)
+            time.sleep(5)
+            for dev in dev_list:
+                unplug_rng(vm, dev)
+
```
Comment (lijinlijin) [inline qemu/tests/rng_hotplug.py]: Why use two different unplug methods here? What's the diff between unplug_all_rngs(vm) and for dev in dev_list: unplug_rng(vm, dev)?
Is there any way to make it one general function?

**Comment by lijinlijin** [comment]
run on win10-32 guest, job can pass:
 (7/9) Host_RHEL.m7.u5.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.rng_hotplug.one_rng.with_migrate_after_unplug: PASS (279.54 s)
 (8/9) Host_RHEL.m7.u5.virtio_rng.rng_random.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.rng_hotplug.one_rng.with_migrate_after_plug: PASS (311.00 s)

File: qemu/tests/rng_hotplug.py [hunk 2]
```diff
@@ -4,23 +4,21 @@
...
 
 @error_context.context_aware
 def run(test, params, env):
     """
     Test hotplug/unplug of rng device
     1) Boot up w/ one rng device
-    2) Unplug rng device
-    3) reboot/shutdown guest(optional)
+    2) Run random read test
+    3) Unplug rng device
```
Comment (lijinlijin) [inline qemu/tests/rng_hotplug.py]: boot guest with one rng device ---> unplug this device, and then repeat hotplug/unplug.
I think you can boot guest without rng device, then start hotplug/unplug loop directly.

Comment (ybduan) [reply qemu/tests/rng_hotplug.py]: The step you mentioned is the original implement of virtio-rng hotplug/unplug. IMHO, it's better to open another automation bug for optimization. Thanks!

File: qemu/tests/rng_hotplug.py (hunk 1 above)
Comment (ybduan) [reply qemu/tests/rng_hotplug.py]: It's temporary workaround for defects of hotplug/unplug/migration functions. Thanks!

**Comment by lijinlijin** [comment]
Acked-by: Li Jin lijin@redhat.com

**Comment by ybduan** [comment]
Could you please help review this pr, thanks! @huangyum

File: qemu/tests/rng_hotplug.py [hunk 3]
```diff
@@ -122,28 +130,36 @@ def run_subtest(sub_test):
                                        {'id': 'virtio-rng-pci-%d' % num})
             hotplug_rng(vm, new_dev)
```
Comment (huangyum) [inline qemu/tests/rng_hotplug.py]: You can invoke simple_hotplug here to support hotplug under q35.  It's ok to send another patch to address this issue, but I'm not sure if below lines would be affected then.

Comment (ybduan) [reply qemu/tests/rng_hotplug.py]: This patch is aiming for migration. So I prefer to investigate what you mentioned above later. What's your opinion? @huangyum  Thanks!

Comment (huangyum) [reply qemu/tests/rng_hotplug.py]: @ybduan I'm fine with that.

**Comment by huangyum** [comment]
LGTM,
Acked-by: Yumei Huang <yuhuang@redhat.com>

File: qemu/tests/rng_hotplug.py [hunk 4]
```diff
@@ -122,28 +130,36 @@ def run_subtest(sub_test):
...
             dev_list.append(new_dev)
+            # temporary workaround for migration
+            vm.params["virtio_rngs"] += ' rng%d' % num
+            vm.params["backend_rng%d" % num] = 'rng-random'
```
Comment (luckyh) [inline qemu/tests/rng_hotplug.py]: I can't see the test restricted other rng types, so should not make this hardcoding.

File: qemu/tests/rng_hotplug.py [hunk 5]
```diff
@@ -122,28 +130,36 @@ def run_subtest(sub_test):
...
+        if sub_test_after_plug == 'migration':
+            # Unplug attached rng device
+            unplug_all_rngs(vm)
+            # temporary workaround for migration
+            vm.params["virtio_rngs"] = ''
+        else:
+            stop_rngd(vm)
+            time.sleep(5)
+            for dev in dev_list:
+                unplug_rng(vm, dev)
```
Comment (luckyh) [inline qemu/tests/rng_hotplug.py]: Why not use the same procedure for both migration and non-migration tests?

**Comment by ybduan** [comment]
The mix of hot-unplugging virtio-rng device and migration is not supported for avocado.
In current code, after hot-unplugging virtio-rng device, the pci address of other devices in the destination will change when doing migration. Migration will fail.
The possible resolutions may be:
1. Remove the default virtio-rng device, add virtio-rng device as user needs in tp-qemu cfg.
2. Optimize migration implementation (destination device's pci address should same as source).

So close this PR for now. Thanks all!
